module/signature/signer_indices.go

```diff
@@ -211,80 +214,88 @@ func EncodeSignersToIndices(
 //  * The input `canonicalIdentifiers` must exhaustively list the set of authorized signers in their canonical order.
 //
 // Expected Error returns during normal operations:
-//  * ErrIncompatibleBitVectorLength indicates that `signerIndices` has the wrong length
-//  * ErrIllegallyPaddedBitVector is the vector is padded with bits other than 0
-//  * ErrInvalidChecksum if the input is shorter than the expected checksum contained therein
+// * signature.InvalidSignerIndicesError if the given index vector `prefixed` does not encode a valid set of signers
 func DecodeSignerIndicesToIdentifiers(
 	canonicalIdentifiers flow.IdentifierList,
 	prefixed []byte,
 ) (flow.IdentifierList, error) {
+	indices, err := decodeSignerIndices(canonicalIdentifiers, prefixed)
+	if err != nil {
+		return nil, err
+	}
+
+	signerIDs := make(flow.IdentifierList, 0, len(indices))
+	for _, index := range indices {
+		signerIDs = append(signerIDs, canonicalIdentifiers[index])
+	}
+	return signerIDs, nil
+}
+
+func decodeSignerIndices(
+	canonicalIdentifiers flow.IdentifierList,
+	prefixed []byte,
+) ([]int, error) {
 	// the prefixed contains the checksum of the canonicalIdentifiers that the signerIndices
 	// creator saw.
 	// extract the checksum and compare with the canonicalIdentifiers to see if both
 	// the signerIndices creator and validator see the same list.
 	signerIndices, err := CompareAndExtract(canonicalIdentifiers, prefixed)
 	if err != nil {
-		return nil, fmt.Errorf("could not extract signer indices from prefixed data: %w", err)
+		if errors.Is(err, ErrInvalidChecksum) {
+			return nil, NewInvalidSignerIndicesErrorf("signer indices' checkum is invalid: %w", err)
+		}
+		return nil, fmt.Errorf("unexpected exception while checking signer indices: %w", err)
 	}
 
 	numberCanonicalNodes := len(canonicalIdentifiers)
 	err = validPadding(signerIndices, numberCanonicalNodes)
 	if err != nil {
-		return nil, fmt.Errorf("signerIndices are invalid: %w", err)
+		if errors.Is(err, ErrIncompatibleBitVectorLength) || errors.Is(err, ErrIllegallyPaddedBitVector) {
+			return nil, NewInvalidSignerIndicesErrorf("invalid padding of signerIndices: %w", err)
```

Comment:
`DecodeSignerIndicesToIdentifiers` and `DecodeSignerIndicesToIdentities` share most of the logic, the code was duplicated. However, I met a bug that here it is supposed to return `NewInvalidSignerIndicesErrorf`, but was returning `NewInvalidSigTypeErrorf`. We fixed it in `DecodeSignerIndicesToIdentifiers`, but not in `DecodeSignerIndicesToIdentities`. 

To prevent this bug from happening, I extracted the common part of the two functions into `decodeSignerIndices`, so that most logic now is shared and reused.